### ISSUE TITLE
IndexOutOfBoundsException with google-storage-wagon 2.0

### ISSUE THREAD

**Issue by rvaidya** [gkatzioura/CloudStorageMaven#44]
Happens every time Maven tries to upload artifacts with version 1.8-2.0.

Using gcloud user credentials (from authentication via CLI), on Windows

**Comment by gkatzioura** [comment]
Thank you for your report! Will look for it and fix it!

**Comment by pbirnie** [comment]
version 1.7 does not produce this exception

**Comment by hadrienk** [comment]
I am getting the same exception:
```
Caused by: java.lang.IndexOutOfBoundsException
    at java.nio.ByteBuffer.wrap (ByteBuffer.java:395)
    at org.eclipse.aether.transport.wagon.WagonTransferListener.transferProgress (WagonTransferListener.java:64)
    at com.gkatzioura.maven.cloud.listener.TransferListenerContainerImpl.lambda$fireTransferProgress$2 (TransferListenerContainerImpl.java:75)
    at java.util.Vector.forEach (Vector.java:1387)
    at com.gkatzioura.maven.cloud.listener.TransferListenerContainerImpl.fireTransferProgress (TransferListenerContainerImpl.java:75)
    at com.gkatzioura.maven.cloud.transfer.TransferProgressImpl.progress (TransferProgressImpl.java:36)
    at com.gkatzioura.maven.cloud.transfer.TransferProgressFileInputStream.read (TransferProgressFileInputStream.java:65)
    at com.vorstella.shade.com.google.api.client.util.ByteStreams.copy (ByteStreams.java:51)
    at com.vorstella.shade.com.google.api.client.util.IOUtils.copy (IOUtils.java:94)
    at com.vorstella.shade.com.google.api.client.http.AbstractInputStreamContent.writeTo (AbstractInputStreamContent.java:72)
    at com.vorstella.shade.com.google.api.client.http.MultipartContent.writeTo (MultipartContent.java:107)
    at com.vorstella.shade.com.google.api.client.http.GZipEncoding.encode (GZipEncoding.java:49)
    at com.vorstella.shade.com.google.api.client.http.HttpEncodingStreamingContent.writeTo (HttpEncodingStreamingContent.java:51)
    at com.vorstella.shade.com.google.api.client.http.javanet.NetHttpRequest$DefaultOutputWriter.write (NetHttpRequest.java:76)
    at com.vorstella.shade.com.google.api.client.http.javanet.NetHttpRequest.writeContentToOutputStream (NetHttpRequest.java:156)
    at com.vorstella.shade.com.google.api.client.http.javanet.NetHttpRequest.execute (NetHttpRequest.java:117)
    at com.vorstella.shade.com.google.api.client.http.javanet.NetHttpRequest.execute (NetHttpRequest.java:84)
    at com.vorstella.shade.com.google.api.client.http.HttpRequest.execute (HttpRequest.java:1011)
    at com.vorstella.shade.com.google.api.client.googleapis.media.MediaHttpUploader.executeCurrentRequestWithoutGZip (MediaHttpUploader.java:548)
    at com.vorstella.shade.com.google.api.client.googleapis.media.MediaHttpUploader.executeCurrentRequest (MediaHttpUploader.java:565)
    at com.vorstella.shade.com.google.api.client.googleapis.media.MediaHttpUploader.directUpload (MediaHttpUploader.java:360)
    at com.vorstella.shade.com.google.api.client.googleapis.media.MediaHttpUploader.upload (MediaHttpUploader.java:334)
    at com.vorstella.shade.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed (AbstractGoogleClientRequest.java:508)
    at com.vorstella.shade.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed (AbstractGoogleClientRequest.java:432)
    at com.vorstella.shade.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute (AbstractGoogleClientRequest.java:549)
    at com.vorstella.shade.com.google.cloud.storage.spi.v1.HttpStorageRpc.create (HttpStorageRpc.java:303)
    at com.vorstella.shade.com.google.cloud.storage.StorageImpl.create (StorageImpl.java:161)
    at com.gkatzioura.maven.cloud.gcs.GoogleStorageRepository.put (GoogleStorageRepository.java:101)
    at com.gkatzioura.maven.cloud.gcs.GoogleStorageWagon.put (GoogleStorageWagon.java:87)
    at org.eclipse.aether.transport.wagon.WagonTransporter$PutTaskRunner.run (WagonTransporter.java:686)
    at org.eclipse.aether.transport.wagon.WagonTransporter.execute (WagonTransporter.java:435)
    at org.eclipse.aether.transport.wagon.WagonTransporter.put (WagonTransporter.java:418)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector$PutTaskRunner.runTask (BasicRepositoryConnector.java:554)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run (BasicRepositoryConnector.java:363)
    at org.eclipse.aether.connector.basic.BasicRepositoryConnector.put (BasicRepositoryConnector.java:287)
    at org.eclipse.aether.internal.impl.DefaultDeployer.deploy (DefaultDeployer.java:295)
    at org.eclipse.aether.internal.impl.DefaultDeployer.deploy (DefaultDeployer.java:211)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.deploy (DefaultRepositorySystem.java:371)
    at org.apache.maven.artifact.deployer.DefaultArtifactDeployer.deploy (DefaultArtifactDeployer.java:142)
    at org.apache.maven.plugin.deploy.AbstractDeployMojo.deploy (AbstractDeployMojo.java:167)
    at org.apache.maven.plugin.deploy.DeployMojo.execute (DeployMojo.java:157)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
```

**Comment by hadrienk** [comment]
Following the stack trace: 
```
 @Override
    public void transferProgress( TransferEvent event, byte[] buffer, int length )
    {
        try
        {
            listener.transportProgressed( ByteBuffer.wrap( buffer, 0, length ) );
        }
        catch ( TransferCancelledException e )
        {
            throw new WagonCancelledException( e );
        }
    }
```

> IndexOutOfBoundsException - If the preconditions on the offset and length parameters do not hold

So I guess a check in TransferListenerContainerImpl should do
```
@Override
    public void fireTransferProgress(Resource resource, int requestType, byte[] buffer, int length) {
        TransferEvent transferEvent = new TransferEvent(this.wagon, resource, TransferEvent.TRANSFER_PROGRESS, requestType);
        transferListeners.forEach(tl->tl.transferProgress(transferEvent,buffer,length));
    }
```

```
    @Override public void progress(byte[] buffer, int length) {
        listenerContainer.fireTransferProgress(this.resource, this.requestType, buffer, Math.min(length, buffer.length));
    }
```

**Comment by gkatzioura** [comment]
It's fixed on 2.3

**Comment by rvaidya** [comment]
@gkatzioura 2.3 is broken because of the issue mentioned in the linked PR.
